### PR TITLE
starts with first internship and company endpoints

### PR DESCRIPTION
**Pull request by LiFaytheGoblin**
add first internship endpoints: get own internship/ by id; get internship by parameters; get internship in certain semester; get internship search options (those that are in the search form so far, but cities and countries are separate; might be interesting for @RobinUndersigned)

add get all companies endpint (might be interesting for @jakepanter)

new: do not return contact info of internship officer - that info is private! instead use company.email or company.website

internships returned to students are not selected at random out of all for which query applies

tested & debugged

Documentation: https://www.getpostman.com/collections/3406cd53e9316e2cf003

One problem remains: search by parameters does not work - something wrong with the options object I construct. I can't find out why.

### CODE REVIEW

**Comment by LiFaytheGoblin** [comment]
Wie hast du das geschafft? WO ist dieser draft button? Ich bin wohl echt zu doof^^

File: server/src/routes/internship.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    "size",
+  ]).toUpperCase(),
+  validate,
+  asyncHandler(findInternships)
+);
+
+internshipRouter.get(
+  "/",
+  authMiddleware(true),
+  query(["semester"]).toUpperCase(),
```
Comment (LiFaytheGoblin) [inline server/src/routes/internship.ts:38]: query("semester")
    .toUpperCase()
    .custom((s) => Semester.isValidSemesterString(s) || !s),

File: server/src/routes/company.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+const companyRouter = Router();
+
+companyRouter.get("/", authMiddleware(), validate, asyncHandler(getAllCompanies));
+
+companyRouter.get("/cities", authMiddleware(), validate, asyncHandler(getCities));
+
+companyRouter.get(
+  "/:country/cities",
+  authMiddleware(),
```
Comment (LiFaytheGoblin) [inline server/src/routes/company.ts:18]: authMiddleware(true)

File: server/src/routes/company.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+import authMiddleware from "../authentication/middleware";
+import { param } from "express-validator";
+import { validate } from "../helpers/validation";
+import * as asyncHandler from "express-async-handler";
+import { Router } from "express";
+import { getAllCompanies, getAllCountries, getCities } from "../controllers/company";
+
+const companyRouter = Router();
+
+companyRouter.get("/", authMiddleware(), validate, asyncHandler(getAllCompanies));
```
Comment (LiFaytheGoblin) [inline server/src/routes/company.ts:12]: authMiddleware(true)

**Comment by Kaes3kuch3n** [comment]
> Wie hast du das geschafft? WO ist dieser draft button? Ich bin wohl echt zu doof^^

Keine Ahnung, der ist irgendwie nicht immer da 😅 Hab ihn gerade zufällig gesehen und einfach mal drauf gedrückt

**Comment by LiFaytheGoblin** [comment]
One problem remains: search by parameters does not work - something wrong with the options object I construct. I can't find out why.

**Comment by Kaes3kuch3n** [comment]
> One problem remains: search by parameters does not work - something wrong with the options object I construct. I can't find out why.

@LiFaytheGoblin Kannst Du dafür nen Issue erstellen und hier verlinken?

**Review by Kaes3kuch3n — APPROVED** [review]

**Comment by LiFaytheGoblin** [comment]
@Kaes3kuch3n #165

**Pull request merged**